rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala

```diff
@@ -0,0 +1,375 @@
+                           rebalanceDelete(currentPath, elems, None)
+                         case _ =>
+                           Applicative[F].pure[List[Trie]](Nil)
+                       }
+            _ <- historyStore.put(elements)
+          } yield elements.lastOption.map(Trie.hash).getOrElse(currentRoot)
+      }
+      .map(newRoot => this.copy(root = newRoot))
+
+  private[history] def traverseTrie(key: KeyPath, start: Blake2b256Hash): F[(Trie, TriePath)] = {
+    type Params = (Trie, List[Byte], TriePath)
+    //t: Trie, currentPath: List[Byte], path: TriePath
+    def traverse(params: Params): F[Either[Params, (Trie, TriePath)]] =
+      params match {
+        case (EmptyTrie, _, path) => Applicative[F].pure((EmptyTrie, path)).map(_.asRight)
+        case (l: Leaf, _, path)   => Applicative[F].pure((l, path append (Nil, l))).map(_.asRight)
+        case (s @ Skip(affix, p), currentPath, path) if currentPath.startsWith(affix.toSeq) =>
+          historyStore
+            .get(p)
+            .map(
+              trie =>
+                (
+                  trie,
+                  currentPath.drop(affix.size.toInt),
+                  path.append(affix.toSeq.toList, s)
+                ).asLeft
+            )
+        case (s: Skip, _, path) => //not a prefix
+          Applicative[F].pure((EmptyTrie, path.copy(conflicting = Some(s)))).map(_.asRight)
+        case (node @ Node(ptr), h :: tail, path) =>
+          pointerBlockStore.get(ptr).flatMap {
+            case Some(pb) =>
+              Applicative[F]
+                .pure((pb.toVector(toInt(h)), tail, path.append(h :: Nil, node)))
+                .map(_.asLeft)
+            case None => Sync[F].raiseError(new RuntimeException("malformed trie"))
+          }
+        case _ => Sync[F].raiseError(new RuntimeException("malformed trie"))
+      }
+    for {
+      hash   <- historyStore.get(start)
+      result <- FlatMap[F].tailRecM((hash, key, TriePath.empty))(traverse)
+    } yield result
+
+  }
+
+  private[history] def traverseTrie(key: KeyPath): F[(Trie, TriePath)] =
+    traverseTrie(key, root)
+}
+
+object History {
+
+  def skipOrFetch[F[_]: Applicative](
+      path: List[Byte],
+      pointer: Blake2b256Hash,
+      fetch: Blake2b256Hash => F[Trie]
+  ): F[Trie] =
+    if (path.isEmpty) {
+      fetch(pointer)
+    } else {
+      Applicative[F].pure(Skip(ByteVector(path), pointer))
+    }
+
+  def skipOrValue(path: List[Byte], t: Trie): Trie =
+    if (path.isEmpty) {
+      t
+    } else {
+      skip(t, path)
+    }
+
+  // this mapping is kept explicit on purpose
+  @inline
+  def toInt(b: Byte): Int =
+    java.lang.Byte.toUnsignedInt(b)
+
+  // this mapping is kept explicit on purpose
+  @inline
+  def toByte(i: Int): Byte =
+    i.toByte
+
+  def commonPrefix(l: List[Byte], r: List[Byte]): List[Byte] =
+    (l.view, r.view).zipped.takeWhile { case (ll, rr) => ll == rr }.unzip._1.toList
```

Comment:
For some reason IDEA suggests `.map(_._1).toList` instead `.unzip.toList`. Guess it saves some allocations?

Reply:
does it? I find it harder to follow.